### PR TITLE
Phase2-hgx109  Complete the geometry parameter part for new HGCal geometry

### PR DESCRIPTION
**Pull request by bsunanda**
Need changes in DetId format to accommodate

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 1]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+#endif
+}
+
+void HGCalGeomParameters::loadSpecParsHexagon8(const DDFilteredView& fv,
+					       HGCalParameters& php) {
+
+  DDsvalues_type sv(fv.mergedSpecifics());
+  php.cellThickness_ = DDVectorGetter::get("CellThickness");
+  for (double & k : php.cellThickness_) 
+    k *= k_ScaleFromDDD;
```
Comment (ianna) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: @bsunanda - this code fragment calls for use of a lambda expression:
std::for_each(php.cellThickness_.begin(), php.cellThickness_.end(), [](int &n){ n*=k_ScaleFromDDD; });

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 2]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+#ifdef EDM_ML_DEBUG
+  edm::LogVerbatim("HGCalGeom") << "HGCalGeomParameters: cell Thickness "
+				<< php.cellThickness_[0] << ":" 
+				<< php.cellThickness_[1] << ":" 
+				<< php.cellThickness_[2];
+#endif
+  php.radius100to200_ = DDVectorGetter::get("Radius100to200");
+  for (double & k : php.radius100to200_)
```
Comment (ianna) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: same here... see the above comments

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 3]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+  edm::LogVerbatim("HGCalGeom") << "HGCalGeomParameters: Polynomial "
+				<< "parameters for 120 to 200 micron "
+				<< "transition "  << php.radius100to200_[0] 
+				<< ":" << php.radius100to200_[1] << ":"
+				<< php.radius100to200_[2] << ":" 
+				<< php.radius100to200_[3] << ":"
+				<< php.radius100to200_[4];
+#endif
+  php.radius200to300_ = DDVectorGetter::get("Radius200to300");
+  for (double & k : php.radius200to300_)
```
Comment (ianna) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: and here...

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 4]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+#endif
+  dummy = DDVectorGetter::get("SlopeBottom");
+  php.slopeMin_   = dummy[0];
+#ifdef EDM_ML_DEBUG
+  edm::LogVerbatim("HGCalGeom") << "HGCalGeomParameters: SlopeBottom " 
+				<< php.slopeMin_;
+#endif
+  php.slopeTop_  = DDVectorGetter::get("SlopeTop");
+  php.zFront_    = DDVectorGetter::get("ZFront");
+  for (double & k : php.zFront_)
```
Comment (ianna) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: and here...

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 5]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+#ifdef EDM_ML_DEBUG
+  edm::LogVerbatim("HGCalGeom") << "HGCalGeomParameters: SlopeBottom " 
+				<< php.slopeMin_;
+#endif
+  php.slopeTop_  = DDVectorGetter::get("SlopeTop");
+  php.zFront_    = DDVectorGetter::get("ZFront");
+  for (double & k : php.zFront_) 
+    k *= k_ScaleFromDDD;
+  php.rMaxFront_ = DDVectorGetter::get("RMaxFront");
+  for (double & k : php.rMaxFront_)
```
Comment (ianna) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: and here...

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 6]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+    k *= k_ScaleFromDDD;
+#ifdef EDM_ML_DEBUG
+  for (unsigned int k = 0; k < php.zFront_.size(); ++k)
+    edm::LogVerbatim("HGCalGeom") << "HGCalParameters: Boundary[" << k 
+				  << "] Z = " << php.zFront_[k] << " Slope = "
+				  << php.slopeTop_[k] << " rMax = "
+				  << php.rMaxFront_[k];
+#endif
+  php.zRanges_   = DDVectorGetter::get("ZRanges");
+  for (double & k : php.zRanges_)
```
Comment (ianna) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: and here...

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 7]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+				<< php.zRanges_[3];
+#endif
+}
+
+void HGCalGeomParameters::loadSpecParsTrapezoid(const DDFilteredView& fv,
+						HGCalParameters& php) {
+
+  DDsvalues_type sv(fv.mergedSpecifics());
+  php.radiusMixBoundary_ = DDVectorGetter::get("RadiusMixBoundary");
+  for (double & k : php.radiusMixBoundary_)
```
Comment (ianna) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: and here...

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 8]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+						HGCalParameters& php) {
+
+  DDsvalues_type sv(fv.mergedSpecifics());
+  php.radiusMixBoundary_ = DDVectorGetter::get("RadiusMixBoundary");
+  for (double & k : php.radiusMixBoundary_) 
+    k *= k_ScaleFromDDD;
+  php.nPhiBinBH_         = dbl_to_int(DDVectorGetter::get("NPhiBinBH"));
+  php.dPhiEta_.clear();
+  for (auto const & nbin : php.nPhiBinBH_) 
+    php.dPhiEta_.emplace_back(CLHEP::twopi/nbin);
```
Comment (ianna) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: @bsunanda - please use M_PI, not CLHEP

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 9]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+#endif
+  std::vector<double> dummy = DDVectorGetter::get("SlopeBottom");
+  php.slopeMin_   = dummy[0];
+#ifdef EDM_ML_DEBUG
+  edm::LogVerbatim("HGCalGeom") << "HGCalGeomParameters: SlopeBottom " 
+				<< php.slopeMin_;
+#endif
+  php.slopeTop_  = DDVectorGetter::get("SlopeTop");
+  php.zFront_    = DDVectorGetter::get("ZFront");
+  for (double & k : php.zFront_)
```
Comment (ianna) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: same comment applies here...

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 10]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+#ifdef EDM_ML_DEBUG
+  edm::LogVerbatim("HGCalGeom") << "HGCalGeomParameters: SlopeBottom " 
+				<< php.slopeMin_;
+#endif
+  php.slopeTop_  = DDVectorGetter::get("SlopeTop");
+  php.zFront_    = DDVectorGetter::get("ZFront");
+  for (double & k : php.zFront_) 
+    k *= k_ScaleFromDDD;
+  php.rMaxFront_ = DDVectorGetter::get("RMaxFront");
+  for (double & k : php.rMaxFront_)
```
Comment (ianna) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: same comment applies here...

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 11]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+    k *= k_ScaleFromDDD;
+#ifdef EDM_ML_DEBUG
+  for (unsigned int k = 0; k < php.zFront_.size(); ++k)
+    edm::LogVerbatim("HGCalGeom") << "HGCalParameters: Boundary[" << k 
+				  << "] Z = " << php.zFront_[k] << " Slope = "
+				  << php.slopeTop_[k] << " rMax = " 
+				  << php.rMaxFront_[k];
+#endif
+  php.zRanges_   = DDVectorGetter::get("ZRanges");
+  for (double & k : php.zRanges_)
```
Comment (ianna) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: same comment applies here...

File: Geometry/HGCalGeometry/src/HGCalGeometryLoader.cc [hunk 12]
```diff
@@ -7,7 +7,7 @@
 #include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
 #include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
 
-//#define EDM_ML_DEBUG
+#define EDM_ML_DEBUG
```
Comment (ianna) [inline Geometry/HGCalGeometry/src/HGCalGeometryLoader.cc]: please, remove it

**Review by ianna — CHANGES_REQUESTED** [review]

File: Geometry/HGCalCommonData/interface/HGCalParameters.h [hunk 13]
```diff
@@ -16,6 +16,7 @@ class HGCalParameters {
 public:
 
   typedef std::vector<std::unordered_map<int32_t,int32_t> > layer_map;
+  typedef std::unordered_map<unsigned int, unsigned int>    wafer_map;
```
[the remark this answers is not in the document]
Comment (kpedro88) [reply Geometry/HGCalCommonData/interface/HGCalParameters.h]: @ianna I do not think that recommendation has universal agreement.

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 14]
```diff
@@ -375,72 +378,227 @@ void HGCalGeomParameters::loadGeometryHexagon(const DDFilteredView& _fv,
...
+ 
+  DDFilteredView fv = _fv;
+  bool dodet(true);
+  std::map<int,HGCalGeomParameters::layerParameters> layers;
+  std::vector<HGCalParameters::hgtrform> trforms;
+  std::vector<bool>                      trformUse;
+  
+  while (dodet) {
+    const DDSolid & sol  = fv.logicalPart().solid();
+    std::string name = sol.name();
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: Cannot use - because sol.name() is doing an operation ns+name()

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 15]
```diff
@@ -375,72 +378,227 @@ void HGCalGeomParameters::loadGeometryHexagon(const DDFilteredView& _fv,
...
+      trformUse.emplace_back(false);
+    }
+    dodet = fv.next();
+  }
+
+  double rmin(0), rmax(0);
+  for (unsigned int i=0; i<layers.size(); ++i) {
+    for (auto & layer : layers) {
+      if (layer.first == (int)(i+1)) {
+	php.layerIndex_.emplace_back(i);
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: This part is a copy of the earlier method

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 16]
```diff
@@ -654,16 +906,149 @@ void HGCalGeomParameters::loadCellParsHexagon(const DDCompactView* cpv,
...
+    edm::LogVerbatim("HGCalGeom") << "[" << k << "]: " << php.cellCoarse_[k];
+#endif
+}
+
+void HGCalGeomParameters::loadSpecParsHexagon8(const DDFilteredView& fv,
+					       HGCalParameters& php) {
+
+  DDsvalues_type sv(fv.mergedSpecifics());
+  php.cellThickness_ = DDVectorGetter::get("CellThickness");
+  for (double & k : php.cellThickness_)
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: removing this causes crash

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 17]
```diff
@@ -375,72 +378,227 @@ void HGCalGeomParameters::loadGeometryHexagon(const DDFilteredView& _fv,
...
+    const DDSolid & sol  = fv.logicalPart().solid();
+    std::string     name = sol.name();
+    // Layers first
+    std::vector<int> copy = fv.copyNumbers();
+    int nsiz = (int)(copy.size());
+    int lay  = (nsiz > 0) ? copy[nsiz-1] : 0;
+    int zp   = (nsiz > 3) ? copy[3] : -1;
+    if (zp != 1) zp = -1;
+    if (lay == 0) {
+      edm::LogError("HGCalGeom") << "Funny layer # " << lay << " zp "
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: this message could be combined with the exception

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 18]
```diff
@@ -375,72 +378,227 @@ void HGCalGeomParameters::loadGeometryHexagon(const DDFilteredView& _fv,
...
+				 << zp << " in " << nsiz << " components";
+      throw cms::Exception("DDException") << "Funny layer # " << lay;
+    } else {
+      if (std::find(php.layer_.begin(),php.layer_.end(),lay) ==
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: how often is this `find()` going to be called, and how large is `php.layer_`? Consider using a more efficient search, if it's possible to sort `php.layer_`.

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 19]
```diff
@@ -375,72 +378,227 @@ void HGCalGeomParameters::loadGeometryHexagon(const DDFilteredView& _fv,
...
+    int zp   = (nsiz > 3) ? copy[3] : -1;
+    if (zp != 1) zp = -1;
+    if (lay == 0) {
+      edm::LogError("HGCalGeom") << "Funny layer # " << lay << " zp "
+				 << zp << " in " << nsiz << " components";
+      throw cms::Exception("DDException") << "Funny layer # " << lay;
+    } else {
+      if (std::find(php.layer_.begin(),php.layer_.end(),lay) == 
+	  php.layer_.end()) php.layer_.emplace_back(lay);
+      std::map<int,HGCalGeomParameters::layerParameters>::iterator itr = layers.find(lay);
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: `auto itr`

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 20]
```diff
@@ -375,72 +378,227 @@ void HGCalGeomParameters::loadGeometryHexagon(const DDFilteredView& _fv,
...
+	layers[lay] = laypar;
+      }
+      DD3Vector x, y, z;
+      fv.rotation().GetComponents( x, y, z ) ;
+      const CLHEP::HepRep3x3 rotation ( x.X(), y.X(), z.X(),
+					x.Y(), y.Y(), z.Y(),
+					x.Z(), y.Z(), z.Z() );
+      const CLHEP::HepRotation hr ( rotation );
+      double xx = k_ScaleFromDDD*fv.translation().X();
+      if (std::abs(xx) < 0.001) xx = 0;
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: make the tolerance value 0.001 into a named constant

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 21]
```diff
@@ -375,72 +378,227 @@ void HGCalGeomParameters::loadGeometryHexagon(const DDFilteredView& _fv,
...
+				  << php.rMinLayHex_[i] << ":" 
+				  << php.rMaxLayHex_[i] << " at z = "  
+				  << php.zLayerHex_[i];
+  }
+  edm::LogVerbatim("HGCalGeom") << "Obtained " << php.trformIndex_.size() 
+				<< " transformation matrices";
+  for (unsigned int k=0; k<php.trformIndex_.size(); ++k) {
+    edm::LogVerbatim("HGCalGeom") << "Matrix[" << k << "] (" << std::hex 
+				  << php.trformIndex_[k]
+				  << std::dec << ") Trnaslation ("
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: typo Translation

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 22]
```diff
@@ -619,22 +783,110 @@ void HGCalGeomParameters::loadWaferHexagon(HGCalParameters& php) {
...
+    }
+  }
+#endif
+}
+
+void HGCalGeomParameters::loadWaferHexagon8(HGCalParameters& php) {
+
+  double waferW(k_ScaleFromDDD*php.waferSize_);
+  double waferS(k_ScaleFromDDD*php.sensorSeparation_);
+  std::unique_ptr<HGCalWaferType> wType =
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: `auto wType`

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 23]
```diff
@@ -619,22 +783,110 @@ void HGCalGeomParameters::loadWaferHexagon(HGCalParameters& php) {
...
+#endif
+  // Clear the vectors
+  php.waferCopy_.clear();
+  php.waferTypeL_.clear();
+  php.waferTypeT_.clear();
+  php.waferPosX_.clear();
+  php.waferPosY_.clear();
+  double r     = 0.5*(waferW+waferS);
+  double dy    = 0.5*r*sqrt(3.0);
+  int    N     = (int)(0.5*rout/r) + 2;
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: check if r = 0?

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc (hunk 17 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: done

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc (hunk 18 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: It does not come in event loop - only during initialization

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc (hunk 18 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: It does not come in event loop - only during initialization

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc (hunk 16 above)
[the remark this answers is not in the document]
Comment (ianna) [reply Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: @bsunanda - you can cherry-pick my commit:
https://github.com/bsunanda/cmssw/commit/2054ca3a61df287cd703953555547a7d6c51e694

BTW, it does improve timing as well:
```
< Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 30-Apr-2018 10:25:04.268 CEST
< %MSG-i HGCalGeom:  HGCalParameterTester:testEE 30-Apr-2018 10:25:04 CEST  Run: 1 Event: 1
---
> Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 30-Apr-2018 10:26:57.447 CEST
> %MSG-i HGCalGeom:  HGCalParameterTester:testEE 30-Apr-2018 10:26:57 CEST  Run: 1 Event: 1
15c15
< %MSG-i HGCalGeom:  HGCalParameterTester:testEE 30-Apr-2018 10:25:04 CEST  Run: 1 Event: 1
---
> %MSG-i HGCalGeom:  HGCalParameterTester:testEE 30-Apr-2018 10:26:58 CEST  Run: 1 Event: 1
1063,1064c1063,1064
< Elapsed time: 0.660254 s
< %MSG-i HGCalGeom:  HGCalParameterTester:testHESil 30-Apr-2018 10:25:04 CEST  Run: 1 Event: 1
---
> Elapsed time: 0.654509 s
> %MSG-i HGCalGeom:  HGCalParameterTester:testHESil 30-Apr-2018 10:26:58 CEST  Run: 1 Event: 1
1067c1067
< %MSG-i HGCalGeom:  HGCalParameterTester:testHESil 30-Apr-2018 10:25:05 CEST  Run: 1 Event: 1
---
> %MSG-i HGCalGeom:  HGCalParameterTester:testHESil 30-Apr-2018 10:26:58 CEST  Run: 1 Event: 1
2145,2146c2145,2146
< Elapsed time: 0.521812 s
< %MSG-i HGCalGeom:  HGCalParameterTester:testHESci 30-Apr-2018 10:25:05 CEST  Run: 1 Event: 1
---
> Elapsed time: 0.519326 s
> %MSG-i HGCalGeom:  HGCalParameterTester:testHESci 30-Apr-2018 10:26:58 CEST  Run: 1 Event: 1
2149c2149
< %MSG-i HGCalGeom:  HGCalParameterTester:testHESci 30-Apr-2018 10:25:05 CEST  Run: 1 Event: 1
---
> %MSG-i HGCalGeom:  HGCalParameterTester:testHESci 30-Apr-2018 10:26:59 CEST  Run: 1 Event: 1
2220c2220
< Elapsed time: 0.518329 s
---
> Elapsed time: 0.515879 s

```

**Comment by bsunanda** [comment]
@ianna @kpedro88 Please approve this PR

**Comment by kpedro88** [comment]
+1

**Comment by ianna** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**